### PR TITLE
ZIP32 key derivation

### LINKED ISSUE #30: Implement ZIP 32 for Orchard
- https://zips.z.cash/zip-0032#specification-orchard-key-derivation
- https://zips.z.cash/zip-0032#orchard-key-path
- https://zips.z.cash/zip-0032#orchard-full-viewing-key-fingerprints-and-tags
- https://zips.z.cash/zip-0032#orchard-extended-spending-keys
  - We won't be exposing Orchard extended spending keys to users by default (unlike Sapling).

### PR DESCRIPTION
**Pull request by therealyingtong**
Closes #30 

Needs test vectors from https://github.com/zcash-hackworks/zcash-test-vectors/issues/41.

### CODE REVIEW

File: src/zip32.rs [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+
+#[allow(non_snake_case)]
+impl ExtendedSpendingKey {
+    /// Returns the spending key of the child key corresponding to
+    /// the path derived from the master key
+    ///
+    /// # Panics
+    ///
+    /// Panics if seed results in invalid spending key.
+    pub fn from_path(seed: &[u8], path: &[u32]) -> Self {
```
Comment (daira) [inline src/zip32.rs:111]: Consider making this return a Result.

File: src/zip32.rs [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+
+/// A hardened child index for a derived key.
+#[derive(Clone, Debug, PartialEq)]
+struct ChildIndex(u32);
+
+impl From<u32> for ChildIndex {
+    /// # Panics
+    ///
+    /// `index` must be less than 2^31
+    fn from(index: u32) -> Self {
```
Comment (daira) [inline src/zip32.rs:69]: Consider making this `try_from`.

File: src/zip32.rs [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+
+    #[test]
+    fn derive_child() {
+        let seed = [0; 32];
+        let xsk_m = ExtendedSpendingKey::master(&seed).unwrap();
+
+        let i_5 = 5;
+        let xsk_5 = xsk_m.derive_child(i_5.into());
+
+        assert!(xsk_5.is_ok());
```
Comment (daira) [inline src/zip32.rs:212]: Check this is a particular value (or replace with more thorough test vectors).

**Review by daira — APPROVED** [review]
utACK modulo comments.

**Review by nuttycom — CHANGES_REQUESTED** [review]
Some minor changes are required to improve API safety.

**Review by nuttycom — APPROVED** [review]
ACK  e4a54cdf6

**Pull request merged**